### PR TITLE
Add expect/forbid context helpers

### PR DESCRIPTION
**Pull request by fregante**
Such utility functions make context-limiting it short and consistent. The functions are still flexible enough to let you customize the error type and/or the error message.

There are some other context exclusions in the code (e.g. multiple contexts allowed) but I left those out for now since they're one or two.

### CODE REVIEW

File: src/background/requests.ts [hunk 1]
```diff
@@ -122,9 +122,7 @@ async function authenticate(
   config: SanitizedServiceConfiguration,
   request: AxiosRequestConfig
 ): Promise<AxiosRequestConfig> {
-  if (!isBackgroundPage()) {
-    throw new Error("authenticate can only be called from the background page");
-  }
```
Comment (fregante) [inline src/background/requests.ts:127]: For unknown reasons, the moment I move this exact piece of code to `util.ts`, the tests fail (as you can see in the first commit in this PR). For this reason I reverted the changes to this file.

**Comment by twschiller** [comment]
> Such utility functions make context-limiting it short and consistent

One thing we'll need to keep our eye on is how easy/hard it is to mock these methods in tests

> I left those out for now since they're one or two.

I think it could make the first parameter take a `Context | Context[]` and then use `castArray`. As is, with them just accepting one context, I think personally think it's cleaner to include the context name in the method name, e.g., `requireBackgroundPage`

> For unknown reasons, the moment I move this exact piece of code

My guess is a mock isn't working properly once you move it. I haven't run the tests, but the sorts of places I'd look for would be: 

* http://github.com/pixiebrix/pixiebrix-extension/blob/b588518a9b109e68346450496ca060bb6bd274ec/src/background/requests.test.ts#L25-L25

File: src/background/requests.ts (hunk 1 above)
Comment (twschiller) [reply src/background/requests.ts]: > For this reason I reverted the changes to this file.

As part of this PR, I would expect us to be able to use expectContext/forbidContext with the tests passing (and remove any instances of isBackgroundPage preconditions)

**Comment by fregante** [comment]
> > Such utility functions make context-limiting it short and consistent
> 
> One thing we'll need to keep our eye on is how easy/hard it is to mock these methods in tests

They're still based on detect-page so we can probably just mock that as usual 👍 

> I think it could make the first parameter take a `Context | Context[]` and then use `castArray`

That's how I was writing it but saw it was basically never used at the moment, so I dropped the complexity.

> I think personally think it's cleaner to include the context name in the method name, e.g., `requireBackgroundPage`

Can do! I was expecting to find more contexts but it ended up just being 2. I think I'll move this to its own `utils/expect-context.ts` file

> My guess is a mock isn't working properly once you move it.

Makes sense! I'll look into it

File: src/utils.ts [hunk 2]
```diff
@@ -355,3 +356,59 @@ export function isPrivatePageError(error: unknown): boolean {
...
+}
+
+/**
+ * @example forbidContext('background')
+ * @example forbidContext('background', WrongContextError)
+ * @example forbidContext('background', 'Wrong context and this is my custom error')
+ * @example forbidContext('background', new Error('Wrong context and this is my custom error'))
+ */
+export function forbidContext(context: Context, error?: ErrorBaseType): void {
+  if (contextsMap.get(context)()) {
```
Comment (twschiller) [inline src/utils.ts:411]: This should throw an error if someone passes in a context that is unexpected

File: src/utils.ts [hunk 3]
```diff
@@ -355,3 +356,59 @@ export function isPrivatePageError(error: unknown): boolean {
...
+ * @example expectContext('background', 'Wrong context and this is my custom error')
+ * @example expectContext('background', new Error('Wrong context and this is my custom error'))
+ */
+export function expectContext(context: Context, error?: ErrorBaseType): void {
+  if (!contextsMap.get(context)()) {
+    throw createError(
+      `This code can only run in the ${context} context`,
+      error
+    );
+  }
```
Comment (twschiller) [inline src/utils.ts:401]: Throw an error if the person passes in a context that doesn't exist in the map. Yes, TypeScript should catch it, but it  is still proper defensive programming to catch it as well

File: src/utils.ts [hunk 4]
```diff
@@ -355,3 +356,59 @@ export function isPrivatePageError(error: unknown): boolean {
...
+  ["background", isBackgroundPage],
+  ["contentScript", isContentScript],
+]);
+
+/**
+ * Accepts 'This is my error' | new Error('This is my error') | Error;
+ * The constructor would be used to create a custom error with the defalt message
+ */
+type ErrorBaseType = string | Error | { new (message?: string): Error };
+function createError(
```
Comment (twschiller) [inline src/utils.ts:373]: This method name is misleading. It's throwing an error, which will break up the control flow. So should be called `throwError`

File: src/utils.ts (hunk 2 above)
Comment (fregante) [reply src/utils.ts:411]: Impossible since this is a TypeScript project

Comment (fregante) [reply src/utils.ts:411]: But I'm converting to `expectBackgroundPage` as you requested, right?

**Review by twschiller — CHANGES_REQUESTED** [review]
Overall, I personally don't think this PR adds much (as the it ends up being approximately the same number of lines for each precondition)

But OK to merge it in as all the pre conditions have been updated to use it

File: src/utils.ts (hunk 4 above)
Comment (fregante) [reply src/utils.ts:373]: `createError` does not throw the error, it only creates one and returns it

Comment (fregante) [reply src/utils.ts:373]: Ouch you're right, that was not the intention

File: src/utils.ts (hunk 2 above)
Comment (twschiller) [reply src/utils.ts:411]: > But I'm converting to expectBackgroundPage as you requested, right?

Personally I'd be in favor of `requireContext(Context | Context[])` or `forbidContext(Context | Context[])`. Or does some of the other predicates, e.g., "isExtensionPage" better cover the use case of code that works on extension pages (options, devtools, panel iframe)

**Comment by fregante** [comment]
> Overall, I personally don't think this PR adds much

One brick doesn't add much, many bricks hold a roof. Many of my PRs have been groundwork for other features or to make the code consistent. Now we can just type `expectBackgroundPage()` in the header of any function without having to copy/paste/customize an error message for each.

This is also giving a change to look into the tests/mocks, so not a bad deal for a minutes’ worth of change 😃

File: src/background/requests.test.ts [hunk 5]
```diff
@@ -19,11 +19,17 @@ import { SanitizedServiceConfiguration, ServiceConfig } from "@/core";
...
 
 const axiosMock = new MockAdapter(axios);
+const mockIsBackgroundPage = isBackgroundPage as jest.MockedFunction<
+  typeof isBackgroundPage
+>;
+mockIsBackgroundPage.mockImplementation(() => false);
 
 jest.mock("@/background/protocol");
 jest.mock("@/auth/token");
+jest.mock("webext-detect-page");
```
Comment (fregante) [inline src/background/requests.test.ts:32]: Typed as suggested in https://jestjs.io/docs/mock-function-api#jestmockedfunction

I think this mocking of `webext-detect-page` can be used in several tests in the future

**Review by twschiller — APPROVED** [review]
Go ahead an merge when ready 👍

**Pull request merged**